KeyVault.Acmebot/Providers/TransIpProvider.cs

```diff
@@ -0,0 +1,333 @@
+        }
+
+        public async Task<IReadOnlyList<DnsZone>> ListZonesAsync()
+        {
+            var zones = await _transIpClient.ListZonesAsync();
+
+            return zones.Select(d => new DnsZone() { Id = d.Name, Name = d.Name }).ToArray();
+        }
+
+        private class TransIpClient
+        {
+            public TransIpClient(string customerName, CryptographyClient cryptoClient)
+            {
+                _customerName = customerName;
+                _cryptoClient = cryptoClient;
+
+                _httpClient = new HttpClient()
+                {
+                    BaseAddress = new Uri("https://api.transip.nl/v6/")
+                };
+
+                _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            }
+
+            private readonly HttpClient _httpClient;
+            private readonly string _customerName;
+            private readonly CryptographyClient _cryptoClient;
+            private TransIPToken _token = null;
+
+            public async Task<IEnumerable<Domain>> ListZonesAsync()
+            {
+                await EnsureLoggedInAsync();
+
+                var response = await _httpClient.GetAsync("domains");
+
+                response.EnsureSuccessStatusCode();
+
+                var domains = await response.Content.ReadAsAsync<ListDomainsResult>();
+
+                return domains.Domains;
+            }
+
+            public async Task<IReadOnlyList<DnsEntry>> ListRecords(string zoneName)
+            {
+                await EnsureLoggedInAsync();
+
+                HttpResponseMessage response = await _httpClient.GetAsync($"domains/{zoneName}/dns");
+
+                response.EnsureSuccessStatusCode();
+
+                ListDnsEntriesResponse entries = await response.Content.ReadAsAsync<ListDnsEntriesResponse>();
+
+                return entries.DnsEntries.ToList();
+            }
+
+            public async Task DeleteRecord(string zoneName, DnsEntry entry)
+            {
+                await EnsureLoggedInAsync();
+
+                var request = new DnsEntryRequest
+                {
+                    DnsEntry = entry
+                };
+
+                var response = await _httpClient.SendAsync(new HttpRequestMessage(HttpMethod.Delete, $"domains/{zoneName}/dns")
+                {
+                    Content = new StringContent(JsonConvert.SerializeObject(request), Encoding.UTF8, "application/json")
+                });
+
+                response.EnsureSuccessStatusCode();
+            }
+
+            public async Task AddRecord(string zoneName, DnsEntry entry)
+            {
+                await EnsureLoggedInAsync();
+
+                var request = new DnsEntryRequest
+                {
+                    DnsEntry = entry
+                };
+
+                var response = await _httpClient.SendAsync(new HttpRequestMessage(HttpMethod.Post, $"domains/{zoneName}/dns")
```

Comment:
Prefer use `PostAsJsonAsync` method.

Reply:
I've tried using PostAsJsonAsync, the problem with that api is that the object is serialized but sent as chunked encoding. The TransIp api requires an explicit content-length header, so this approach does not work unfortunately.